presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java

```diff
@@ -599,6 +588,56 @@ public Optional<PlanNode> visitProject(ProjectNode node, Context context)
         return Optional.of(canonicalPlan);
     }
 
+    // Variable names and plan node ids can change with what order we process nodes because of our
+    // stateful canonicalization using `variableAllocator` and `planNodeIdAllocator`.
+    // We want to order sources in a consistent manner, because the order matters when hashing plan.
+    // Returns a list of indices in input sources array, with a canonical order.
+    private Optional<List<Integer>> orderSources(List<PlanNode> sources)
+    {
+        // Try heuristic where we sort sources by the tables they scan.
+        Optional<List<Integer>> sourcesByTables = orderSourcesByTables(sources);
+        if (sourcesByTables.isPresent()) {
+            return sourcesByTables;
+        }
+
+        // We canonicalize each source independently, and use its representation to order sources.
+        Multimap<String, Integer> sourceToPosition = TreeMultimap.create();
+        for (int i = 0; i < sources.size(); ++i) {
+            Optional<CanonicalPlan> canonicalSource = generateCanonicalPlan(sources.get(i), strategy, objectMapper);
+            if (!canonicalSource.isPresent()) {
+                return Optional.empty();
+            }
+            sourceToPosition.put(canonicalSource.get().toString(objectMapper), i);
+        }
+        return Optional.of(sourceToPosition.values().stream().collect(toImmutableList()));
+    }
+
+    // Order sources by list of tables they use. If any 2 sources are using the same set of tables, we give up
+    // and return Optional.empty().
+    // Returns a list of indices in input sources array, with a canonical order
+    private Optional<List<Integer>> orderSourcesByTables(List<PlanNode> sources)
+    {
+        Multimap<String, Integer> sourceToPosition = TreeMultimap.create();
+        for (int i = 0; i < sources.size(); ++i) {
+            List<String> tables = new ArrayList<>();
+            forTree(PlanNode::getSources).depthFirstPreOrder(sources.get(i)).forEach(node -> {
```

Comment:
is there a reason for using this instead of PlanNodeSearcher findAll()

Reply:
Just didnt know about it